File: python/02-two-pointers/009-container-with-most-water.py
"""
NeetCode 150 - Container With Most Water

Approach:
Use two pointers starting from both ends.
At each step, compute the area formed by the two lines:
area = (right - left) * min(height[left], height[right]).
Move the pointer at the smaller height inward,
since the smaller height limits the area.

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List

class Solution:
    def maxArea(self, heights: List[int]) -> int:
        
        left, right = 0, len(heights) - 1
        bestArea = 0

        while left < right:
            currArea = (right - left) * min(heights[left], heights[right])
            bestArea = max(bestArea, currArea)

            if heights[left] <= heights[right]:
                left += 1
            else:
                right -= 1

        return bestArea

File: java/02-two-pointers/009-container-with-most-water.java
/*
 * NeetCode 150 - Container With Most Water
 *
 * Approach:
 * Use two pointers (left and right). Compute area formed by the two lines:
 * area = (right - left) * min(height[left], height[right]).
 * Move the pointer at the smaller height inward, since the smaller height
 * limits the area; moving the taller pointer cannot increase the min height.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

class Solution {
    public int maxArea(int[] height) {
        int left = 0, right = height.length - 1;
        int best = 0;

        while (left < right) {
            int h = Math.min(height[left], height[right]);
            int area = (right - left) * h;
            best = Math.max(best, area);

            if (height[left] <= height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return best;
    }
}
